File: constructs/pulumi/csharp/Nautilus.Infra.Pulumi.Tests/ComponentsTest.cs
using System.Collections.Immutable;
using Pulumi.Testing;
using Xunit;
using Nautilus.Infra.Pulumi;
using Nautilus.Infra.Pulumi.Policy;

namespace Nautilus.Infra.Pulumi.Tests;

/// <summary>
/// Component unit tests — C#
/// ==========================
/// Uses Pulumi mock infrastructure to verify component wiring
/// without deploying any real Azure resources.
///
/// Run:
///   dotnet test Nautilus.Infra.Pulumi.Tests/
/// </summary>
public class ComponentsTest
{
    // ── Tagging ────────────────────────────────────────────────────────────────

    [Fact]
    public void RequiredTags_ContainsMandatoryKeys()
    {
        var tags = Tagging.RequiredTags("myapp", "dev");
        Assert.Equal("pulumi",  tags["managed_by"]);
        Assert.Equal("myapp",   tags["project"]);
        Assert.Equal("dev",     tags["environment"]);
    }

    [Fact]
    public void RequiredTags_OverridesExtraKey()
    {
        var extra = new Dictionary<string, string>
        {
            ["managed_by"] = "manual",
            ["team"]       = "platform",
        };
        var tags = Tagging.RequiredTags("myapp", "prod", extra);
        Assert.Equal("pulumi",   tags["managed_by"]);
        Assert.Equal("platform", tags["team"]);
    }


    // ── NetworkComponent ───────────────────────────────────────────────────────

    [Fact]
    public void Network_ThrowsOnInvalidEnvironment()
    {
        Assert.Throws<ArgumentException>(() =>
            new NetworkComponent("net", new NetworkComponentArgs(
                Project:       "myapp",
                Environment:   "uat",
                ResourceGroup: "rg",
                Location:      "eastus",
                AddressSpace:  ["10.0.0.0/16"])));
    }

    [Fact]
    public async Task Network_VnetIdIsOutput()
    {
        var resources = await Testing.RunAsync<NetworkComponentStack>();
        var vnet = resources.OfType<global::Pulumi.AzureNative.Network.VirtualNetwork>().FirstOrDefault();
        Assert.NotNull(vnet);
    }

    [Fact]
    public void Network_SubnetDelegationConstructed()
    {
        var deleg = new SubnetDelegation(
            Name:    "postgres",
            Service: "Microsoft.DBforPostgreSQL/flexibleServers",
            Actions: ["Microsoft.Network/virtualNetworks/subnets/join/action"]);
        Assert.Equal("Microsoft.DBforPostgreSQL/flexibleServers", deleg.Service);
    }


    // ── DatabaseComponent ──────────────────────────────────────────────────────

    [Fact]
    public void Database_ThrowsOnInvalidEnvironment()
    {
        Assert.Throws<ArgumentException>(() =>
            new DatabaseComponent("db", new DatabaseComponentArgs(
                Project:       "myapp",
                Environment:   "qa",
                ResourceGroup: "rg",
                Location:      "eastus",
                SubnetId:      "sn",
                DnsZoneId:     "dns",
                AdminPassword: "secret")));
    }

    [Fact]
    public void Database_HaEnabledConstructed()
    {
        // Verify the record is created correctly with HA enabled — resource-level
        // validation happens at deployment time.
        var cfg = new PostgresConfig(
            Databases: ["appdb"],
            HaEnabled: true);
        Assert.True(cfg.HaEnabled);
    }

    [Fact]
    public void Database_DefaultConfigHaDisabled()
    {
        var cfg = new PostgresConfig([]);
        Assert.False(cfg.HaEnabled);
        Assert.Equal(32, cfg.StorageGb);
        Assert.Equal("15", cfg.PgVersion);
    }


    // ── AksComponent ──────────────────────────────────────────────────────────

    [Fact]
    public void Aks_ThrowsOnInvalidEnvironment()
    {
        Assert.Throws<ArgumentException>(() =>
            new AksComponent("aks", new AksComponentArgs(
                Project:        "myapp",
                Environment:    "badenv",
                ResourceGroup:  "rg",
                Location:       "eastus",
                SubnetId:       "sn",
                LogWorkspaceId: "ws")));
    }

    [Fact]
    public void Aks_DefaultConfigValues()
    {
        var cfg = new AksConfig();
        Assert.Equal("1.29",              cfg.KubernetesVersion);
        Assert.Equal("Standard_D2s_v3",   cfg.SystemNodeVmSize);
        Assert.Equal(3,                   cfg.SystemNodeCount);
        Assert.Equal("10.240.0.0/16",     cfg.ServiceCidr);
        Assert.Equal("10.240.0.10",       cfg.DnsServiceIp);
    }

    [Fact]
    public void Aks_NodePoolConfigDefaults()
    {
        var pool = new NodePoolConfig();
        Assert.Equal("Standard_D4s_v3", pool.VmSize);
        Assert.Equal(2,  pool.NodeCount);
        Assert.False(pool.EnableAutoScaling);
        Assert.Equal(1,  pool.MinCount);
        Assert.Equal(10, pool.MaxCount);
    }
}

// ── Minimal stack for Testing.RunAsync ────────────────────────────────────────

public class NetworkComponentStack : global::Pulumi.Stack
{
    public NetworkComponentStack()
    {
        new NetworkComponent("net", new NetworkComponentArgs(
            Project:       "myapp",
            Environment:   "dev",
            ResourceGroup: "myapp-dev-rg",
            Location:      "eastus",
            AddressSpace:  ["10.10.0.0/16"],
            Subnets: new Dictionary<string, SubnetConfig>
            {
                ["aks"] = new SubnetConfig(AddressPrefix: "10.10.0.0/22"),
            }));
    }
}

File: constructs/pulumi/java/src/test/java/com/nautilus/infra/pulumi/ComponentsTest.java
package com.nautilus.infra.pulumi;

import com.nautilus.infra.pulumi.policy.Tagging;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Component unit tests — Java
 * ============================
 * Validates component configuration wiring and tag policy.
 * Tests that exercise runtime resource creation would require a Pulumi mock
 * provider; validation tests here focus on construction-time checks which
 * do not require Azure credentials.
 *
 * Run:
 *   mvn test
 */
class ComponentsTest {

    // ── Tagging ────────────────────────────────────────────────────────────────

    @Test
    void tagging_requiredTagsContainsMandatoryKeys() {
        var tags = Tagging.requiredTags("myapp", "dev");
        assertEquals("pulumi",  tags.get("managed_by"));
        assertEquals("myapp",   tags.get("project"));
        assertEquals("dev",     tags.get("environment"));
    }

    @Test
    void tagging_requiredTagsOverrideExtra() {
        var extra = Map.of("managed_by", "manual", "team", "platform");
        var tags  = Tagging.requiredTags("myapp", "prod", extra);
        assertEquals("pulumi",   tags.get("managed_by"),
            "required tag managed_by must override extra");
        assertEquals("platform", tags.get("team"));
    }

    @Test
    void tagging_extraKeysIncluded() {
        var tags = Tagging.requiredTags("svc", "staging", Map.of("cost_center", "eng"));
        assertEquals("eng", tags.get("cost_center"));
    }


    // ── NetworkComponent props and validation ──────────────────────────────────

    @Test
    void network_throwsOnInvalidEnvironment() {
        assertThrows(IllegalArgumentException.class, () ->
            new NetworkComponent("net",
                NetworkComponent.NetworkComponentProps.builder()
                    .project("myapp").environment("uat")
                    .resourceGroup("rg").location("eastus")
                    .addressSpace(List.of("10.0.0.0/16"))
                    .build(),
                null));
    }

    @Test
    void network_subnetDelegationBuilder() {
        var deleg = NetworkComponent.SubnetDelegation.builder()
            .name("postgres")
            .service("Microsoft.DBforPostgreSQL/flexibleServers")
            .actions(List.of("Microsoft.Network/virtualNetworks/subnets/join/action"))
            .build();
        assertEquals("Microsoft.DBforPostgreSQL/flexibleServers", deleg.getService());
        assertEquals(1, deleg.getActions().size());
    }

    @Test
    void network_subnetConfigBuilder() {
        var cfg = NetworkComponent.SubnetConfig.builder()
            .addressPrefix("10.0.0.0/24")
            .serviceEndpoints(List.of("Microsoft.ContainerRegistry"))
            .build();
        assertEquals("10.0.0.0/24", cfg.getAddressPrefix());
        assertEquals(1, cfg.getServiceEndpoints().size());
    }


    // ── DatabaseComponent props and validation ─────────────────────────────────

    @Test
    void database_throwsOnInvalidEnvironment() {
        assertThrows(IllegalArgumentException.class, () ->
            new DatabaseComponent("db",
                DatabaseComponent.DatabaseComponentProps.builder()
                    .project("myapp").environment("qa")
                    .resourceGroup("rg").location("eastus")
                    .subnetId("sn").dnsZoneId("dns")
                    .adminPassword("secret")
                    .build(),
                null));
    }

    @Test
    void database_defaultConfigValues() {
        var cfg = DatabaseComponent.PostgresConfig.builder().build();
        assertFalse(cfg.isHaEnabled());
        assertEquals("GP_Standard_D2s_v3", cfg.getSku());
        assertEquals(32, cfg.getStorageGb());
        assertEquals("15", cfg.getPgVersion());
        assertFalse(cfg.isGeoRedundant());
    }

    @Test
    void database_haEnabledFlagSet() {
        var cfg = DatabaseComponent.PostgresConfig.builder()
            .haEnabled(true)
            .databases(List.of("appdb"))
            .build();
        assertTrue(cfg.isHaEnabled());
        assertTrue(cfg.getDatabases().contains("appdb"));
    }


    // ── AksComponent props and validation ─────────────────────────────────────

    @Test
    void aks_throwsOnInvalidEnvironment() {
        assertThrows(IllegalArgumentException.class, () ->
            new AksComponent("aks",
                AksComponent.AksComponentProps.builder()
                    .project("myapp").environment("badenv")
                    .resourceGroup("rg").location("eastus")
                    .subnetId("sn").logWorkspaceId("ws")
                    .build(),
                null));
    }

    @Test
    void aks_defaultConfigValues() {
        var cfg = AksComponent.AksConfig.builder().build();
        assertEquals("1.29",            cfg.getKubernetesVersion());
        assertEquals("Standard_D2s_v3", cfg.getSystemNodeVmSize());
        assertEquals(3,                 cfg.getSystemNodeCount());
        assertEquals("10.240.0.0/16",   cfg.getServiceCidr());
        assertEquals("10.240.0.10",     cfg.getDnsServiceIp());
    }

    @Test
    void aks_nodePoolConfigDefaults() {
        var pool = AksComponent.NodePoolConfig.builder().build();
        assertEquals("Standard_D4s_v3", pool.getVmSize());
        assertEquals(2,  pool.getNodeCount());
        assertFalse(pool.isEnableAutoScaling());
        assertEquals(1,  pool.getMinCount());
        assertEquals(10, pool.getMaxCount());
    }

    @Test
    void aks_additionalNodePoolBuilder() {
        var pool = AksComponent.NodePoolConfig.builder()
            .vmSize("Standard_D8s_v3")
            .enableAutoScaling(true)
            .minCount(2).maxCount(10)
            .labels(Map.of("workload", "app"))
            .build();
        assertEquals("Standard_D8s_v3", pool.getVmSize());
        assertTrue(pool.isEnableAutoScaling());
        assertEquals(2, pool.getMinCount());
        assertEquals("app", pool.getLabels().get("workload"));
    }
}
